scvi/inference/inference.py

```diff
@@ -71,3 +80,221 @@ def train(self, n_path=10, n_epochs=50, **kwargs):
+    trainer_specific_kwargs: dict = None,
+    train_func_specific_kwargs: dict = None,
+    use_batches: bool = False,
+    verbose: bool = True,
+    is_best_training: bool = False,
+) -> Union[Dict[str, Any], Type[VAE]]:
+    """Objective function for automatic hyperparameter optimization.
+    Train a scVI model and return the best value of the early-stopping metric (e.g, log-likelihood).
+
+    :param space: dict containing up to three sub-dicts with keys 'model_tunable_kwargs',
+    'trainer_tunable_kwargs' or 'train_func_tunable_kwargs'.
+    Each of those dict contains hyperopt defined parameter spaces (e.g. ``hp.choice(..)``)
+    which will be passed to the corresponding object : model, trainer or train method
+    when performing hyperoptimization.
+    :param gene_dataset: scVI gene dataset
+    :param model_class: scVI model class (e.g ``VAE``, ``VAEC``, ``SCANVI``)
+    :param trainer_class: Trainer class (e.g ``UnsupervisedTrainer``)
+    :param model_specific_kwargs: dict of fixed parameters which will be passed to the model.
+    :param trainer_specific_kwargs: dict of fixed parameters which will be passed to the trainer.
+    :param train_func_specific_kwargs: dict of fixed parameters which will be passed to the train method.
+    :param use_batches: If False, pass n_batch=0 to model else pass gene_dataset.n_batches
+    :param is_best_training: True if training the model with the best hyperparameters
+    :return: best value of the early stopping metric, and best model if is_best_training
+    """
+    # hyperopt params
+    space = defaultdict(dict, space)
+    model_tunable_kwargs = space['model_tunable_kwargs']
+    trainer_tunable_kwargs = space['trainer_tunable_kwargs']
+    train_func_tunable_kwargs = space['train_func_tunable_kwargs']
+
+    if verbose and not is_best_training:
+        print('Parameters being tested: ')
+        print('model:')
+        print(model_tunable_kwargs)
+        print('trainer: ')
+        print(trainer_tunable_kwargs)
+        print('train method: ')
+        print(train_func_tunable_kwargs)
+
+    # define model
+    model_specific_kwargs.update(model_tunable_kwargs)
+    model = model_class(
+        n_input=gene_dataset.nb_genes,
+        n_batch=gene_dataset.n_batches * use_batches,
+        **model_specific_kwargs,
+    )
+
+    # add hardcoded parameters
+    if is_best_training:
+        trainer_specific_kwargs['train_size'] = 1.0
+        # no monitoring, will crash otherwise
+        trainer_specific_kwargs['frequency'] = None
+    else:
+        # evaluate at each epoch
+        trainer_specific_kwargs['frequency'] = 1
+
+    # define trainer
+    trainer_specific_kwargs.update(trainer_tunable_kwargs)
+    trainer = trainer_class(
+        model,
+        gene_dataset,
+        **trainer_specific_kwargs,
+    )
+
+    # redirect scVI progress bar to null because of incompatibility with hyperopt
+    f = open(os.devnull, 'w')
+    orig_stdout = sys.stdout
+    sys.stdout = f
```

Comment:
It's clever you thought of this, but if it's just to avoid errors from Jupyter notebook, should we make this redirect something we do in the Jupyter notebook rather than in the function?

Reply:
The bad behaviour also happens when running in script
![image](https://user-images.githubusercontent.com/23578633/56462179-a4df5780-6373-11e9-9d05-9da465ba56e3.png)

Reply:
Looks like an issue with tqdm, which both scVI and hyperopt have as dependencies. It isn't supposed to work that way. It may also have to do with how the terminal executible was compiled. Let me search a bit for bug reports.

Reply:
Do you have the latest version of the tqdm installed? If not, would you try manually installing the latest one to see if the undesirable behavior persists? It might be that some dependency is keeping the latest version of tqdm from getting installed. I see some bug reports online that suggest this issue with tqdm may have been fixed.

Reply:
It's still there

![image](https://user-images.githubusercontent.com/23578633/56462920-0efdf980-6380-11e9-8e83-1b155b3cfe62.png)

The issue is documented in hyperopt and might cause them to revert their progbar apparently: https://github.com/hyperopt/hyperopt/issues/476

Reply:
A possible workaround would be to make scVI's progress bar optional
It seems like the best option right now

Reply:
That seems like a good solution